File: agent/app/schemas/report.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

ConfidenceFlag = Literal["extracted", "inferred", "confirmed"]


class MedicationOrder(BaseModel):
    drug_name: str
    dose: str
    frequency: str
    duration: str
    route: str | None = None


class FollowUp(BaseModel):
    needed: bool
    timeframe: str | None = None
    reason: str | None = None


class Subjective(BaseModel):
    chief_complaint: str
    history_of_present_illness: str
    symptom_duration: str | None = None
    associated_symptoms: list[str] = Field(default_factory=list)
    relevant_history: list[str] = Field(default_factory=list)


class Objective(BaseModel):
    vital_signs: dict[str, str] = Field(default_factory=dict)
    physical_exam: str | None = None


class Assessment(BaseModel):
    primary_diagnosis: str
    differential_diagnoses: list[str] = Field(default_factory=list)
    icd10_codes: list[str] = Field(default_factory=list)


class Plan(BaseModel):
    medications: list[MedicationOrder] = Field(default_factory=list)
    investigations: list[str] = Field(default_factory=list)
    lifestyle_advice: list[str] = Field(default_factory=list)
    follow_up: FollowUp
    red_flags: list[str] = Field(default_factory=list)


class MedicalReport(BaseModel):
    subjective: Subjective
    objective: Objective = Field(default_factory=Objective)
    assessment: Assessment
    plan: Plan
    confidence_flags: dict[str, ConfidenceFlag] = Field(default_factory=dict)


def required_field_is_missing(report: MedicalReport) -> str | None:
    """Return the first required field that is blank/empty, or None if complete."""
    if not report.subjective.chief_complaint.strip():
        return "subjective.chief_complaint"
    if not report.subjective.history_of_present_illness.strip():
        return "subjective.history_of_present_illness"
    if not report.assessment.primary_diagnosis.strip():
        return "assessment.primary_diagnosis"
    for med in report.plan.medications:
        for attr in ("drug_name", "dose", "frequency", "duration"):
            if not getattr(med, attr).strip():
                return "plan.medications"
    if report.plan.follow_up.needed and not (report.plan.follow_up.timeframe or "").strip():
        return "plan.follow_up.needed"
    return None

File: frontend/lib/types/report.ts
// frontend/lib/types/report.ts
// Mirror of agent/app/schemas/report.py::MedicalReport and
// backend/.../MedicalReportDto.java. Field names are camelCase at this layer
// (the backend remaps snake_case to camelCase in its DTO).
//
// See spec §4.7 — single source of truth. Do not drift.

export type ConfidenceFlag = "extracted" | "inferred" | "confirmed";

export interface MedicationOrder {
  drugName: string;
  dose: string;
  frequency: string;
  duration: string;
  route?: string | null;
}

export interface FollowUp {
  needed: boolean;
  timeframe?: string | null;
  reason?: string | null;
}

export interface Subjective {
  chiefComplaint: string;
  historyOfPresentIllness: string;
  symptomDuration?: string | null;
  associatedSymptoms: string[];
  relevantHistory: string[];
}

export interface Objective {
  vitalSigns: Record<string, string>;
  physicalExam?: string | null;
}

export interface Assessment {
  primaryDiagnosis: string;
  differentialDiagnoses: string[];
  icd10Codes: string[];
}

export interface Plan {
  medications: MedicationOrder[];
  investigations: string[];
  lifestyleAdvice: string[];
  followUp: FollowUp;
  redFlags: string[];
}

export interface MedicalReport {
  subjective: Subjective;
  objective: Objective;
  assessment: Assessment;
  plan: Plan;
  confidenceFlags: Record<string, ConfidenceFlag>;
}

export interface Clarification {
  field: string;
  prompt: string;
  context: string;
}

export type ReviewStatus = "complete" | "clarification_pending" | "error";

export interface ReportReviewResult {
  status: ReviewStatus;
  report: MedicalReport | null;
  clarification: Clarification | null;
}

export interface ChatTurn {
  turnIndex: number;
  role: "user" | "assistant";
  content: string;
  toolCallName?: string | null;
  createdAt?: string | null;
}
